src/com/google/enterprise/adaptor/DocIdSender.java

```diff
@@ -211,16 +216,78 @@ public DocId pushNamedResources(Map<DocId, Acl> resources,
+   * feed replaces, rather than augments or updates, all groups from a
+   * given data source. This is useful because many connectors cannot
+   * detect when a group has been deleted, so repeated incremental updates
+   * would leave deleted groups in the GSA's groups database.
+   *
+   * When pushing a full feed, we must do it in a single "batch", because if
+   * done in batches and a later batch fails (or it simply takes a non-trivial
+   * amount of time to upload all the batches), we would end up with an
+   * incomplete set of groups.
+   *
+   * The problem with supplying all the group definitions in a single full
+   * feed is the GSA limit to the size of a feed file (1GB). Although this
+   * is a large limit, several of our customers would easily exceed that.
+   * One customer has 100,000 groups averaging 10,000 members each. Another
+   * has 400,000 groups.
+   *
+   * A novel solution to this is to double-buffer the group definitions for
+   * each data source. We maintain two pseudo data sources for each actual
+   * data source (*-A and *-B). Each "full" upload alternates between these
+   * two, using batches of incremental feeds. As each batch is processed,
+   * groups defined in that batch are re-assigned from the older data source
+   * to the newer one. Once all batches have been uploaded, the only groups
+   * remaining in the older data source are the ones that were missing from
+   * the newer feeds (in other words, deleted from the repository). We then
+   * "delete" these stale entries by sending an empty, but true full feed
+   * for the older data source. After this, the GSA should only have group
+   * definitions for the newer data source.
+   *
+   * If there is a failure mid-push, the GSA groups database may have entries
+   * for both data sources. Since the adaptors do not maintain persistent
+   * state, on a restart it will not know which is "older" vs. "newer".
+   * This is not fatal however, as the only downside might be some deleted
+   * groups persisting through one or two "full" pushes before they get
+   * cleaned up.
+   */
+  private Map<String, String> groupSources = new HashMap<String, String>();
+
+  /**
+   * Returns the previous group source pushed to. If there was no previous
+   * group source, returns the supplied source. If a connector only ever does
+   * INCREMENTAL feeds, then that group source will always be used. This is
+   * also backward compatible. If a connector is upgraded to a new version that
+   * supports FULL feeds, then the older incremental group source will
+   * eventually get deleted from the GSA.
+   */
+  private String previousGroupSource(String source) {
+    String prevSource = groupSources.get(source);
+    return (prevSource == null) ? source : prevSource;
+  }
+
+  /** Returns the alternate to the previous group source. */
+  private String alternateGroupSource(String source) {
+    return (source + "-A").equals(groupSources.get(source))
```

Comment:
These names are admin-visible in the admin console. Is there something more intuitive? I think numbers might be slightly more explicable, in that splitting foo into foo-1 and foo-2 might be more idiomatic than foo-A and foo-B. Maybe even -full-1 and -full-2? The datasource name length is 256, so we're not stealing characters from a particularly rare resource (I thought the limit was 32, for example).

For posterity, I didn't like the use of 3 possible name sources, but I came to like the fact that the use of the decorated names indicates that a full feed was used at some point. Might be useful info for debugging.